src/promptflow/promptflow/executor/_async_nodes_scheduler.py

```diff
@@ -0,0 +1,79 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+import asyncio
+from typing import Dict, List, Tuple, Any
+
+from promptflow._core.flow_execution_context import FlowExecutionContext
+from promptflow._core.tools_manager import ToolsManager
+from promptflow._utils.logger_utils import flow_logger
+from promptflow.contracts.flow import Node
+from promptflow.executor._dag_manager import DAGManager
+from promptflow.executor._errors import NoNodeExecutedError
+
+
+class AsyncNodesScheduler:
+    def __init__(
+        self,
+        tools_manager: ToolsManager,
+        inputs: Dict,
+        nodes: List[Node],
+        node_concurrency: int,
+        context: FlowExecutionContext,
+    ) -> None:
+        self._tools_manager = tools_manager
+        self._node_concurrency = node_concurrency
+        flow_logger.info(f"Start to run {len(nodes)} nodes with the current event loop.")
+        self._dag_manager = DAGManager(nodes, inputs)
+        self._context = context
+        self._task2node: Dict[Any, Node] = {}
+
+    async def execute(
+        self,
+    ) -> Tuple[dict, dict]:
+        loop = asyncio.get_event_loop()
+        self._execute_nodes(loop)
+        while not self._dag_manager.completed():
+            if not self._task2node:
+                raise NoNodeExecutedError("No nodes are ready for execution, but the flow is not completed.")
+            tasks = [task for task in self._task2node]
+            done, _ = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+            self._dag_manager.complete_nodes({
+                self._task2node[task].name: task.result()
+                for task in done
+            })
+            for task in done:
+                del self._task2node[task]
+            self._execute_nodes(loop)
+        for node in self._dag_manager.bypassed_nodes:
+            self._dag_manager.completed_nodes_outputs[node] = None
+        return self._dag_manager.completed_nodes_outputs, self._dag_manager.bypassed_nodes
+
+    def _execute_nodes(self, loop: asyncio.AbstractEventLoop):
+        # Skip nodes and update node run info until there are no nodes to bypass
+        nodes_to_bypass = self._dag_manager.pop_bypassable_nodes()
+        while nodes_to_bypass:
+            self._bypass_nodes(nodes_to_bypass)
+            nodes_to_bypass = self._dag_manager.pop_bypassable_nodes()
+
+        # Submit nodes that are ready to run
+        nodes_to_exec = self._dag_manager.pop_ready_nodes()
+        tasks = []
+        for each_node in nodes_to_exec:
+            tasks.append(self._create_node_task(loop, each_node))
+        return tasks
```

Comment:
Why return `task` here? It seems that its return value is not used.

Reply:
Refined the logic, you may confirm again.